stream/src/main/java/org/apache/kafka/streaming/processor/TopologyBuilder.java

```diff
@@ -0,0 +1,184 @@
+            this.metadata = metadata;
+        }
+    }
+
+    private class SourceClazz {
+        public Deserializer keyDeserializer;
+        public Deserializer valDeserializer;
+
+        private SourceClazz(Deserializer keyDeserializer, Deserializer valDeserializer) {
+            this.keyDeserializer = keyDeserializer;
+            this.valDeserializer = valDeserializer;
+        }
+    }
+
+    private class SinkClazz {
+        public Serializer keySerializer;
+        public Serializer valSerializer;
+
+        private SinkClazz(Serializer keySerializer, Serializer valSerializer) {
+            this.keySerializer = keySerializer;
+            this.valSerializer = valSerializer;
+        }
+    }
+
+    public TopologyBuilder() {}
+
+    @SuppressWarnings("unchecked")
+    public final void addSource(String name, Deserializer keyDeserializer, Deserializer valDeserializer, String... topics) {
+        for (String topic : topics) {
+            if (topicsToSourceNames.containsKey(topic))
+                throw new IllegalArgumentException("Topic " + topic + " has already been registered by another processor.");
+
+            topicsToSourceNames.put(topic, name);
+        }
+
+        sourceClasses.put(name, new SourceClazz(keyDeserializer, valDeserializer));
+    }
+
+    public final void addSink(String name, Serializer keySerializer, Serializer valSerializer, String... topics) {
+        for (String topic : topics) {
+            if (topicsToSinkNames.containsKey(topic))
+                throw new IllegalArgumentException("Topic " + topic + " has already been registered by another processor.");
+
+            topicsToSinkNames.put(topic, name);
+        }
+
+        sinkClasses.put(name, new SinkClazz(keySerializer, valSerializer));
+    }
+
+    public final void addProcessor(String name, Class<? extends KafkaProcessor> processorClass, ProcessorMetadata config, String... parentNames) {
+        if (processorClasses.containsKey(name))
+            throw new IllegalArgumentException("Processor " + name + " is already added.");
+
+        processorClasses.put(name, new ProcessorClazz(processorClass, config));
+
+        if (parentNames != null) {
+            for (String parent : parentNames) {
+                if (!processorClasses.containsKey(parent))
+                    throw new IllegalArgumentException("Parent processor " + parent + " is not added yet.");
+
+                // add to parent list
+                if (!parents.containsKey(name))
+                    parents.put(name, new ArrayList<>());
+                parents.get(name).add(parent);
+
+                // add to children list
+                if (!children.containsKey(parent))
+                    children.put(parent, new ArrayList<>());
+                children.get(parent).add(name);
+            }
+        }
+    }
+
+    /**
+     * Build the topology by creating the processors
+     */
+    @SuppressWarnings("unchecked")
+    public ProcessorTopology build() {
```

Comment:
Ack on checking uniqueness of the names, with that, there should be no cycles since nodes can only be add once, right?